gtsam/basis/basis.i

```diff
@@ -131,13 +126,6 @@ virtual class ManifoldEvaluationFactor : gtsam::NoiseModelFactor {
 // `ComponentDerivativeFactor`
 
 #include <gtsam/basis/FitBasis.h>
-// We'll allow transparent binding of python dict to Sequence in this
-// compilation unit using pybind11/stl.h.
-// Another alternative would be making Sequence opaque in
-// python/gtsam/{preamble, specializations}, but std::map<double, double> is
-// common enough that it may cause collisions, and we don't need
-// reference-access anyway.
-#include <pybind11/stl.h>
 
```

Comment:
(See also `specializations/basis.h`)

I should have made a python unit test in #832 , but can you give a MWE of how to use FitBasis in python?  I can't figure out how (before this commit, it was just `gtsam.FitBasisChebyshev2(2, {0:0, 1:1, 0.5:0.5}, noiseModel).parameters()`).

(I don't think it is possible with the version you have now; I get this error:
```python
>>> gtsam.Sequence({1:2})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. gtsam.gtsam.Sequence()

Invoked with: {1: 2}
```

I think you also need the requisite line in the preamble, right?:
```c++
PYBIND11_MAKE_OPAQUE(std::map<double, double>);
```
)
But then the syntax for using e.g. FitBasis is a little bit cumbersome:
```python
data = gtsam.Sequence()
for k, v in {0:0, 0.5:0.5, 1:1}.items():
    data[k] = v
gtsam.FitBasisChebyshev2(2, data, noiseModel).parameters()
# syntax with <pybind11/stl.h>:
# gtsam.FitBasisChebyshev2(2, {0:0, 1:1, 0.5:0.5}, noiseModel).parameters()
```
(the constructor `gtsam.Sequence({...})` doesn't exist)

---

By the way, although not an immediate issue, this also has the side-effect of causing binding collisions.  For example, if we ever had another type:
```c++
py::bind_map<gtsam::Sequence>(m_, "Sequence");
py::bind_map<std::map<double, double>>(m_, "AnotherMapDouble");
```
this will throw the runtime python error:
```
>>> import gtsam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gerry/miniconda3/lib/python3.8/site-packages/gtsam-4.1.0-py3.8.egg/gtsam/__init__.py", line 2, in <module>
    from .gtsam import *
ImportError: generic_type: type "AnotherMapDouble" is already registered!
```

This collision occurs if we have any other `map<double, double>` in the entire gtsam namespace (even outside the basis module, e.g. in the geometry module).  In contrast, including `pybind11/stl.h` as I had done before stays within the compilation unit, i.e. it doesn't leave basis.i and we don't get any binding collisions.

Sorry for the long comment but I had already spent a pretty considerable time thinking about / testing these considerations when I made #832 and had pretty intentionally chosen to do it this way, so we could have maybe saved time if we had discussed this together :)

Reply:
The issue with the stl.h include is that it now pulls in that file for all languages, such as Matlab. Did you try testing that with the matlab wrapper? Or a setup where someone only needs Matlab but no python wrapping?

Reply:
I'll also double check the binding collisions, in which case this change is no bueno.

Reply:
Oh shoot sorry I forgot about the matlab wrapper.  My mistake!!!

Binding collisions still applies tho.  I'd be ok calling this a later problem, since currently there's nothing else trying to bind to `std::map<double, double>` (build for now, not for the future)

Not sure but I think maybe placing Sequence inside FitBasis's scope, or inside a different namespace might fix binding collisions???  Something about pybind11 `py::module_local()` but I didn't read in detail: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#binding-stl-containers

Reply:
Nah I want to finish this now. A later problem is not a good enough excuse for me.

Reply:
Great find on `py::module_local`. That should do the trick.

> This module-local binding is designed to avoid potential conflicts between module bindings (for example, from two separate modules each attempting to bind std::vector<int> as a python type).
 
> It is possible to override this behavior to force a definition to be either module-local or global. To do so, you can pass the attributes `py::module_local()` (**to make the binding module-local**) or py::module_local(false) (to make the binding global) into the py::bind_vector or py::bind_map arguments

Reply:
Uhh guess not

Reply:
I am lost, what is your issue here?

Reply:
@ProfFan 
Trying to bind two different `std::map<double, double>` with
```c++
py::bind_map<std::map<double, double>>(m_, "TypeName1");
py::bind_map<std::map<double, double>>(m_, "TypeName2");
```
will cause a python runtime error on import.  This occurs even if the types are bound in different interface files.

I think changing `m_` to a "more local module" would fix this, if possible, but I haven't played with it.  It may also just not be possible to do what we want.

Reply:
module_local is per-module, if we are building several modules (.so) then it should work.

Reply:
I figured it out!!!

Now you can do `coefficients = gtsam.FitBasisChebyshev2(2, {0:0, 1:1, 0.5:0.5}, noiseModel).parameters()` no problem.

Reply:
Unfortunately, the module level thing doesn't work since we are linking all the separate .so files into one. By module, I believe they mean individual files that are not linked together, e.g.

```
gtsam
	- base
		- __init__.py
		- base.so
	- basis
		- __init__.py
		- basis.so
	- linear
		- __init__.py
		- linear.so
...
```

Reply:
Opening a PR where I added the <pybind11/stl.h> include to ".../preamble/basis.h" and everything seems to work fine.

This is a separate PR but I think:
1) as Varun discovered, the bindings for `IndexPairXXX` should be updated/fixed at some point
2) binding collisions is actually not something we should be worried about because we shouldn't really have more than one type describing a `map<double, double>` anyway, i.e. just opaque-ify and bind std::map<double, double>

Reply:
I think the best way is to alias **IN Python**. Just create an alias in the Python files and nothing to worry, yay.

Reply:
@ProfFan can you please explain more?

I came up with an alternative, where in the wrapper definition, we use `std::map` directly rather than `gtsam::Sequence`, so Pybind11 can wrap to a dict easily.